test/wal_reader/wal_reader_test.cpp

```diff
@@ -0,0 +1,298 @@
+
+		{
+			// Read write connection
+			Connection con(db);
+
+			// Execute all test case setup SQL
+			for (const auto &test_case : test_cases) {
+				REQUIRE_NO_FAIL(con.Query(test_case.setup_sql));
+			}
+		}
+
+		// Create a WAL reader to read the WAL file
+		auto file_system = FileSystem::CreateLocal();
+		auto handle = file_system->OpenFile(wal_path, FileFlags::FILE_FLAGS_READ);
+		auto &db_manager = DatabaseManager::Get(*db.instance.get());
+		auto databases = db_manager.GetDatabases(*read_con.context);
+		auto &attached_db = databases[0];
+
+		auto file_reader = make_uniq<BufferedFileReader>(FileSystem::Get(attached_db), std::move(handle));
+		WALReader reader(attached_db, std::move(file_reader));
+
+		// Read and verify WAL entries
+		std::unique_ptr<ParseInfo> entry;
+		vector<bool> found(test_cases.size(), false);
+
+		while ((entry = reader.Next()) != nullptr) {
+			// Try each test case's validator
+			for (size_t i = 0; i < test_cases.size(); i++) {
+				if (test_cases[i].validator(entry.get())) {
+					found[i] = true;
+					break;
+				}
+			}
+		}
+
+		// Verify all test cases were found
+		for (size_t i = 0; i < test_cases.size(); i++) {
+			INFO("Missing WAL record for: " << test_cases[i].name);
+			REQUIRE(found[i]);
+		}
+
+		// Abort the read-only transaction and close the second connection
+		REQUIRE_NO_FAIL(read_con.Query("ROLLBACK"));
+	}
+}
+
+TEST_CASE("Test WAL reader error handling", "[wal_reader]") {
+	string dbdir = TestCreatePath("wal_reader_test_errors");
+	{
+		DuckDB db(dbdir);
+		// Create a second connection for long-running read transaction
+		Connection read_con(db);
+		// Start a read-only transaction in the second connection
+		REQUIRE_NO_FAIL(read_con.Query("BEGIN TRANSACTION READ ONLY"));
+
+		// get the path to db and append ".wal" to it
+		auto wal_path = dbdir + WAL_FILE_EXTENSION;
+
+		{
+			// Read write connection
+			Connection con(db);
+
+			// Create some WAL entries
+			REQUIRE_NO_FAIL(con.Query("CREATE TABLE test (id INTEGER)"));
+		}
+
+		{
+			auto &db_manager = DatabaseManager::Get(*db.instance.get());
+			auto databases = db_manager.GetDatabases(*read_con.context);
+			auto &attached_db = databases[0];
+
+			// Test reading from an empty WAL file
+			auto file_system = FileSystem::CreateLocal();
+			auto empty_wal_path = dbdir + "_empty" + WAL_FILE_EXTENSION;
+			auto handle = file_system->OpenFile(empty_wal_path,
+			                                    FileFlags::FILE_FLAGS_FILE_CREATE_NEW | FileFlags::FILE_FLAGS_WRITE);
+			auto empty_handle = file_system->OpenFile(empty_wal_path, FileFlags::FILE_FLAGS_READ);
+			auto empty_file_reader =
+			    make_uniq<BufferedFileReader>(FileSystem::Get(attached_db), std::move(empty_handle));
+			WALReader empty_reader(attached_db, std::move(empty_file_reader));
+			REQUIRE(empty_reader.Next() == nullptr);
+
+			// Test reading with invalid file handle
+			REQUIRE_THROWS_AS(WALReader(attached_db, nullptr), Exception);
```

Comment:
I don't think you want to use this, in tests we run with "abort on internal exception" configuration, this causes an internal exception, which aborts.
I suggest to just remove this line.